From: Mike Christie <mchristi@redhat.com>
Date: Fri, 5 Mar 2010 02:22:01 -0500
Subject: [scsi] scsi_dh_emc: fix mode select setup
Message-id: <1267755721-8400-1-git-send-email-mchristi@redhat.com>
Patchwork-id: 23496
O-Subject: [PATCH RHEL 6.0]: scsi_dh_emc: fix mode select setup
Bugzilla: 570685
RH-Acked-by: Tomas Henzl <thenzl@redhat.com>
RH-Acked-by: David Milburn <dmilburn@redhat.com>

From: Mike Christie <mchristi@redhat.com>

This is for BZ: 570685.

The patch fixes:
- setting of the length for mode selects.
- setting of the data direction for mode select 10.

This patch is upstream here:
http://git.kernel.org/?p=linux/kernel/git/jejb/scsi-misc-2.6.git;a=commit;h=6c71dcb28ff9b63b814a0b76a256f5dae08d3e0d

EMC had tested the patch when we added this for RHEL 5 and when
I sent upstream. It is not already in RHEL6 because there was a
mixup upstream and it is just now getting merged.

Signed-off-by: Aristeu Rozanski <arozansk@redhat.com>

diff --git a/drivers/scsi/device_handler/scsi_dh_emc.c b/drivers/scsi/device_handler/scsi_dh_emc.c
index 6196675..63032ec 100644
--- a/drivers/scsi/device_handler/scsi_dh_emc.c
+++ b/drivers/scsi/device_handler/scsi_dh_emc.c
@@ -272,7 +272,7 @@ static struct request *get_req(struct scsi_device *sdev, int cmd,
 	int len = 0;
 
 	rq = blk_get_request(sdev->request_queue,
-			(cmd == MODE_SELECT) ? WRITE : READ, GFP_NOIO);
+			(cmd != INQUIRY) ? WRITE : READ, GFP_NOIO);
 	if (!rq) {
 		sdev_printk(KERN_INFO, sdev, "get_req: blk_get_request failed");
 		return NULL;
@@ -286,14 +286,17 @@ static struct request *get_req(struct scsi_device *sdev, int cmd,
 		len = sizeof(short_trespass);
 		rq->cmd_flags |= REQ_RW;
 		rq->cmd[1] = 0x10;
+		rq->cmd[4] = len;
 		break;
 	case MODE_SELECT_10:
 		len = sizeof(long_trespass);
 		rq->cmd_flags |= REQ_RW;
 		rq->cmd[1] = 0x10;
+		rq->cmd[8] = len;
 		break;
 	case INQUIRY:
 		len = CLARIION_BUFFER_SIZE;
+		rq->cmd[4] = len;
 		memset(buffer, 0, len);
 		break;
 	default:
@@ -301,7 +304,6 @@ static struct request *get_req(struct scsi_device *sdev, int cmd,
 		break;
 	}
 
-	rq->cmd[4] = len;
 	rq->cmd_type = REQ_TYPE_BLOCK_PC;
 	rq->cmd_flags |= REQ_FAILFAST_DEV | REQ_FAILFAST_TRANSPORT |
 			 REQ_FAILFAST_DRIVER;
